<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>How to run arraylib interactively &mdash; arraylib 0.1 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="How to run arraylib-solve on the command line" href="run_test_data_cmdline.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> arraylib
          </a>
              <div class="version">
                0.1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../LibraryExperiment.html">LibraryExperiment</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">How to run arraylib interactively</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Read-trimming-and-genomic-alignment">Read trimming and genomic alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Assembly-of-count-matrix-and-preprocessing">Assembly of count matrix and preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Deconvolution-and-location-inference">Deconvolution and location inference</a></li>
<li class="toctree-l2"><a class="reference internal" href="#References">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="run_test_data_cmdline.html">How to run <code class="docutils literal notranslate"><span class="pre">arraylib-solve</span></code> on the command line</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">arraylib</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>How to run arraylib interactively</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/notebooks/run_test_data_interactive.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="How-to-run-arraylib-interactively">
<h1>How to run arraylib interactively<a class="headerlink" href="#How-to-run-arraylib-interactively" title="Permalink to this headline"></a></h1>
<p>First we import the LibraryExperiment class from the package.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from arraylib.libraryexperiment import LibraryExperiment
<br/></pre></div>
</div>
</div>
<p>We generate an experiment object by instantiating the LibraryExperiment class with our chosen parameters.</p>
<p>Required parameters:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">input_dir</span></code>: path to directory holding the input fastq files</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">exp_design</span></code>: path to csv file indicating experimental design (values should be separated by a comma). The experimental design file should have columns, Filename, Poolname and Pooldimension. (see example in tests/test_data/full_exp_design.csv)</p>
<ul>
<li><p>Filename should contain all the unqiue input fastq filenames.</p></li>
<li><p>Poolname should indicate to which pool a given file belongs. Multiple files per poolname are allowed.</p></li>
<li><p>Pooldimension indicates the pooling dimension a pool belongs to. All pools sharing the same pooling dimension should have the same string in the Pooldimension column.</p></li>
</ul>
</li>
</ul>
<p>An example of how an exp_design file could look like:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 41%" />
<col style="width: 24%" />
<col style="width: 35%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Filename</p></th>
<th class="head"><p>Poolname</p></th>
<th class="head"><p>Pooldimension</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>column1.fastq</p></td>
<td><p>column1</p></td>
<td><p>columns</p></td>
</tr>
<tr class="row-odd"><td><p>column2.fastq</p></td>
<td><p>column2</p></td>
<td><p>columns</p></td>
</tr>
<tr class="row-even"><td><p>row1.fastq</p></td>
<td><p>row1</p></td>
<td><p>rows</p></td>
</tr>
<tr class="row-odd"><td><p>row2.fastq</p></td>
<td><p>row2</p></td>
<td><p>rows</p></td>
</tr>
<tr class="row-even"><td><p>platerow1.fastq</p></td>
<td><p>platerow1</p></td>
<td><p>platerows</p></td>
</tr>
<tr class="row-odd"><td><p>platerow2.fastq</p></td>
<td><p>platerow2</p></td>
<td><p>platerows</p></td>
</tr>
<tr class="row-even"><td><p>platecol1.fastq</p></td>
<td><p>platecol1</p></td>
<td><p>platecols</p></td>
</tr>
<tr class="row-odd"><td><p>platecol2.fastq</p></td>
<td><p>platecol2</p></td>
<td><p>platecols</p></td>
</tr>
</tbody>
</table>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gb_ref</span></code> path to genbank reference file</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bowtie_ref</span></code> path to bowtie index files, ending with the basename of your index (if the basename of your index is UTI89 and you store your bowtie2 <a class="reference external" href="#1">[1]</a> references in <code class="docutils literal notranslate"><span class="pre">bowtie_ref</span></code> it should be bowtie_ref/UTI89). Please visit <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-build-indexer">https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-build-indexer</a> for a manual how to create bowtie2 indices.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tn_seq</span></code> transposon sequence (e.g. AGATGTGTATAAGAGACAG)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bar_upstream</span></code> upstream sequence of barcode (e.g. CGAGGTCTCT)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bar_downstream</span></code> downstream sequence of barcode (e.g. CGTACGCTGC)</p></li>
</ul>
<p>Optional parameters:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">map_quality</span></code> minimum bowtie2 alignment quality score for each base to include read</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">seq_quality</span></code> minimum phred score for each base to include read</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">transposon_mismatches</span></code> number of transposon mismatches allowed</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">filter_thr</span></code> threshold for local filter (e.g. a threshold of 0.05 would filter out all reads &lt; 0.05 of the maximum read count for a given mutant)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">global_thr</span></code> threshold for global filter (all reads below g_thr will be set to 0)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">min_counts</span></code> minimum counts of a barcode to be included in analysis</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">use_barcodes</span></code> whether to perform deconvolution only based on barcodes without genomic alignment</p></li>
</ul>
<p>You can also refer to the API to get a detailed description of the input parameter choices.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>experiment = LibraryExperiment(cores=4,
                               map_quality=30,
                               seq_quality=10,
                               gb_ref=&quot;gb_ref&quot;,
                               bowtie_ref=&quot;bowtie_ref/UTI89&quot;,
                               tn_seq=&quot;AGATGTGTATAAGAGACAG&quot;,
                               tn_mismatches=2,
                               input_dir=&quot;input&quot;,
                               exp_design=&quot;full_exp_design.csv&quot;,
                               use_barcodes=False,
                               bar_upstream=&quot;CGAGGTCTCT&quot;,
                               bar_downstream=&quot;CGTACGCTGC&quot;,
                               filter_thr=0.05,
                               global_filter_thr=5,
                               min_counts=5)
<br/></pre></div>
</div>
</div>
<section id="Read-trimming-and-genomic-alignment">
<h2>Read trimming and genomic alignment<a class="headerlink" href="#Read-trimming-and-genomic-alignment" title="Permalink to this headline"></a></h2>
<p>First we detect the transposon sequence in the reads in our input fastq files in <code class="docutils literal notranslate"><span class="pre">input_dir</span></code>. By running:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># detect transposon and trim sequences
experiment.get_genomic_seq()
</pre></div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">get_genomic_seq()</span></code> first detects the transposon sequence in each read, trims the genomic part of the read and saves it to a compiled fastq file (trimmed_sequences.fastq) in a temp directory in the current working directory.</p>
<p>Then we use bowtie2 to align the trimmed reads to the reference genome in <code class="docutils literal notranslate"><span class="pre">bowtie_ref</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># align trimmed sequences to reference using bowtie2
experiment.align_genomic_seq()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
624483 reads; of these:
  624483 (100.00%) were unpaired; of these:
    0 (0.00%) aligned 0 times
    624483 (100.00%) aligned exactly 1 time
    0 (0.00%) aligned &gt;1 times
100.00% overall alignment rate
</pre></div></div>
</div>
</section>
<section id="Assembly-of-count-matrix-and-preprocessing">
<h2>Assembly of count matrix and preprocessing<a class="headerlink" href="#Assembly-of-count-matrix-and-preprocessing" title="Permalink to this headline"></a></h2>
<p>Once the alignment is finished we can count the reads of each mutant in each pool and assemble the count matrix using <code class="docutils literal notranslate"><span class="pre">write_count_matrix()</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># assemble count matrix from aligned reads
experiment.write_count_matrix()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Generating count matrix with 7000 mutants
Generated count matrix!
</pre></div></div>
</div>
<p>Our experiment has now 4 different count matrices stored as attributes:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">raw_count_mat</span></code>: stores the raw reads per mutant and pool</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">count_mat</span></code>: stores the raw reads per mutant and pool, but spurious barcodes are filtered out</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">normalized_count_mat</span></code>: stores the reads from count_mat, but normalized to counts per million</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">filtered_count_mat</span></code>: stores normalized reads, but has filters applied according to <code class="docutils literal notranslate"><span class="pre">filter_thr</span></code> and <code class="docutils literal notranslate"><span class="pre">global_thr</span></code></p></li>
</ul>
<p>In practice it is beneficial to apply filters to the normalized count matrix, as often spurious reads are detected. This usually makes the deconvolution and location inference much harder, as many mutants become ambiguous. Therefore, we usually want to continue with the <code class="docutils literal notranslate"><span class="pre">filtered_count_mat</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>experiment.count_mat
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Feature</th>
      <th>Orientation</th>
      <th>Barcode</th>
      <th>Reference</th>
      <th>1</th>
      <th>10</th>
      <th>11</th>
      <th>12</th>
      <th>2</th>
      <th>3</th>
      <th>...</th>
      <th>PC07</th>
      <th>PC08</th>
      <th>PR01</th>
      <th>PR02</th>
      <th>PR03</th>
      <th>PR04</th>
      <th>PR05</th>
      <th>PR06</th>
      <th>PR07</th>
      <th>PR08</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>49</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0</td>
      <td>0</td>
      <td>10</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>14</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>50</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>13</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>13</td>
    </tr>
    <tr>
      <th>2</th>
      <td>51</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>11</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>53</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>13</td>
      <td>0</td>
      <td>...</td>
      <td>10</td>
      <td>0</td>
      <td>10</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>54</td>
      <td>+</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0</td>
      <td>13</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>14</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>14</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>6995</th>
      <td>7045</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0</td>
      <td>10</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>14</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>13</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>11</td>
      <td>0</td>
    </tr>
    <tr>
      <th>6996</th>
      <td>7046</td>
      <td>+</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0</td>
      <td>10</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>13</td>
      <td>0</td>
      <td>0</td>
      <td>12</td>
      <td>11</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>6997</th>
      <td>7047</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>14</td>
      <td>12</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>11</td>
      <td>0</td>
    </tr>
    <tr>
      <th>6998</th>
      <td>7048</td>
      <td>+</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>14</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>10</td>
      <td>11</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>13</td>
      <td>0</td>
      <td>0</td>
      <td>14</td>
    </tr>
    <tr>
      <th>6999</th>
      <td>7050</td>
      <td>+</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0</td>
      <td>0</td>
      <td>10</td>
      <td>0</td>
      <td>14</td>
      <td>0</td>
      <td>...</td>
      <td>12</td>
      <td>0</td>
      <td>14</td>
      <td>0</td>
      <td>0</td>
      <td>11</td>
      <td>13</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>7000 rows × 40 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>experiment.filtered_count_mat
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Feature</th>
      <th>Orientation</th>
      <th>Barcode</th>
      <th>Reference</th>
      <th>1</th>
      <th>10</th>
      <th>11</th>
      <th>12</th>
      <th>2</th>
      <th>3</th>
      <th>...</th>
      <th>PC07</th>
      <th>PC08</th>
      <th>PR01</th>
      <th>PR02</th>
      <th>PR03</th>
      <th>PR04</th>
      <th>PR05</th>
      <th>PR06</th>
      <th>PR07</th>
      <th>PR08</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>49</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>734.322221</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>732.945919</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>50</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.00000</td>
      <td>675.465032</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>664.349959</td>
    </tr>
    <tr>
      <th>2</th>
      <td>51</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>581.856652</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>3</th>
      <td>53</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>980.022616</td>
      <td>0.000000</td>
      <td>...</td>
      <td>496.15480</td>
      <td>0.000000</td>
      <td>509.735957</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>4</th>
      <td>54</td>
      <td>+</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0.00000</td>
      <td>981.798958</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.00000</td>
      <td>727.423880</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>721.277692</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>6995</th>
      <td>7045</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0.00000</td>
      <td>755.229968</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>1003.152766</td>
      <td>...</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>662.656744</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>549.862534</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>6996</th>
      <td>7046</td>
      <td>+</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0.00000</td>
      <td>755.229968</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>680.592639</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>629.954328</td>
      <td>566.718187</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>6997</th>
      <td>7047</td>
      <td>-</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>732.945919</td>
      <td>635.930048</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>549.862534</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>6998</th>
      <td>7048</td>
      <td>+</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>1108.64745</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>753.863551</td>
      <td>788.191459</td>
      <td>...</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>682.450522</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>715.453802</td>
    </tr>
    <tr>
      <th>6999</th>
      <td>7050</td>
      <td>+</td>
      <td>nan</td>
      <td>NC_007946.1</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>734.322221</td>
      <td>0.0</td>
      <td>1055.408971</td>
      <td>0.000000</td>
      <td>...</td>
      <td>595.38576</td>
      <td>0.000000</td>
      <td>713.630339</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>581.856652</td>
      <td>682.450522</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
<p>7000 rows × 40 columns</p>
</div></div>
</div>
</section>
<section id="Deconvolution-and-location-inference">
<h2>Deconvolution and location inference<a class="headerlink" href="#Deconvolution-and-location-inference" title="Permalink to this headline"></a></h2>
<p>We then use the count matrix to infer the most likely location of each mutant using <code class="docutils literal notranslate"><span class="pre">deconvolve()</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Inferring most likely locations and writing location summary
experiment.deconvolve()
</pre></div>
</div>
</div>
<p>We can access the deconvolution results with <code class="docutils literal notranslate"><span class="pre">location_summary</span></code> to get a detailed information about each mutant detected. To get information about each well in the arrayed call <code class="docutils literal notranslate"><span class="pre">transposed_location_summary</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>experiment.location_summary
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Reference</th>
      <th>Coordinate</th>
      <th>Gene</th>
      <th>Locus_tag</th>
      <th>Predicted_wells</th>
      <th>Gene_orientation</th>
      <th>Transposon_orientation</th>
      <th>Percent_from_start</th>
      <th>Gene_hit</th>
      <th>Ambiguity</th>
      <th>Possible_wells</th>
      <th>Pool_counts</th>
      <th>Barcode</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>NC_007946.1</td>
      <td>49</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>F_11_PC02_PR02</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>False</td>
      <td>Unambiguous</td>
      <td>F_11_PC02_PR02</td>
      <td>11:10.0;F:12.0;PC02:12.0;PR02:14.0</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>1</th>
      <td>NC_007946.1</td>
      <td>50</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>F_5_PC08_PR08</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>False</td>
      <td>Unambiguous</td>
      <td>F_5_PC08_PR08</td>
      <td>5:13.0;F:10.0;PC08:13.0;PR08:13.0</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>2</th>
      <td>NC_007946.1</td>
      <td>51</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>D_8_PC05_PR04</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>False</td>
      <td>Unambiguous</td>
      <td>D_8_PC05_PR04</td>
      <td>8:10.0;D:10.0;PC05:10.0;PR04:11.0</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>3</th>
      <td>NC_007946.1</td>
      <td>53</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>E_2_PC07_PR01</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>False</td>
      <td>Unambiguous</td>
      <td>E_2_PC07_PR01</td>
      <td>2:13.0;E:11.0;PC07:10.0;PR01:10.0</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>4</th>
      <td>NC_007946.1</td>
      <td>54</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>D_10_PC08_PR06</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>False</td>
      <td>Unambiguous</td>
      <td>D_10_PC08_PR06</td>
      <td>10:13.0;D:10.0;PC08:14.0;PR06:14.0</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>6995</th>
      <td>NC_007946.1</td>
      <td>7045</td>
      <td>Not_found</td>
      <td>UTI89_RS00040</td>
      <td>B_10_PC03_PR01;D_3_PC05_PR07;F_7_PC03_PR01</td>
      <td>-</td>
      <td>-</td>
      <td>0.55905</td>
      <td>True</td>
      <td>Ambiguous</td>
      <td>B_10_PC03_PR01;B_10_PC03_PR07;B_10_PC05_PR01;B...</td>
      <td>10:10.0;3:14.0;7:13.0;B:12.0;D:12.0;F:12.0;PC0...</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>6996</th>
      <td>NC_007946.1</td>
      <td>7046</td>
      <td>Not_found</td>
      <td>UTI89_RS00040</td>
      <td>B_10_PC05_PR06;C_6_PC03_PR02;G_4_PC05_PR05</td>
      <td>-</td>
      <td>+</td>
      <td>0.558351</td>
      <td>True</td>
      <td>Ambiguous</td>
      <td>B_10_PC03_PR02;B_10_PC03_PR05;B_10_PC03_PR06;B...</td>
      <td>10:10.0;4:13.0;6:14.0;B:10.0;C:14.0;G:11.0;PC0...</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>6997</th>
      <td>NC_007946.1</td>
      <td>7047</td>
      <td>Not_found</td>
      <td>UTI89_RS00040</td>
      <td>A_6_PC01_PR02;C_9_PC04_PR03;D_8_PC04_PR07</td>
      <td>-</td>
      <td>-</td>
      <td>0.557652</td>
      <td>True</td>
      <td>Ambiguous</td>
      <td>A_6_PC01_PR02;A_6_PC01_PR03;A_6_PC01_PR07;A_6_...</td>
      <td>6:14.0;8:10.0;9:12.0;A:13.0;C:13.0;D:10.0;PC01...</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>6998</th>
      <td>NC_007946.1</td>
      <td>7048</td>
      <td>Not_found</td>
      <td>UTI89_RS00040</td>
      <td>A_1_PC04_PR08;A_2_PC03_PR05;F_3_PC03_PR05</td>
      <td>-</td>
      <td>+</td>
      <td>0.556953</td>
      <td>True</td>
      <td>Ambiguous</td>
      <td>A_1_PC03_PR05;A_1_PC03_PR08;A_1_PC04_PR05;A_1_...</td>
      <td>1:14.0;2:10.0;3:11.0;A:13.0;F:10.0;PC03:10.0;P...</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>6999</th>
      <td>NC_007946.1</td>
      <td>7050</td>
      <td>Not_found</td>
      <td>UTI89_RS00040</td>
      <td>A_5_PC07_PR05;D_11_PC07_PR04;H_2_PC02_PR01</td>
      <td>-</td>
      <td>+</td>
      <td>0.555556</td>
      <td>True</td>
      <td>Ambiguous</td>
      <td>A_11_PC02_PR01;A_11_PC02_PR04;A_11_PC02_PR05;A...</td>
      <td>11:10.0;2:14.0;5:14.0;A:13.0;D:11.0;H:13.0;PC0...</td>
      <td>nan</td>
    </tr>
  </tbody>
</table>
<p>7000 rows × 13 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span>experiment.transposed_location_summary
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Well</th>
      <th>row</th>
      <th>col</th>
      <th>PC</th>
      <th>PR</th>
      <th>Reference</th>
      <th>Coordinate</th>
      <th>Gene</th>
      <th>Locus_tag</th>
      <th>Gene_orientation</th>
      <th>Transposon_orientation</th>
      <th>Percent_from_start</th>
      <th>Gene_hit</th>
      <th>Ambiguity</th>
      <th>Barcode</th>
    </tr>
    <tr>
      <th>Well</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A_1_PC01_PR01</th>
      <td>A_1_PC01_PR01</td>
      <td>A</td>
      <td>1</td>
      <td>PC01</td>
      <td>PR01</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>A_1_PC01_PR02</th>
      <td>A_1_PC01_PR02</td>
      <td>A</td>
      <td>1</td>
      <td>PC01</td>
      <td>PR02</td>
      <td>NC_007946.1;NC_007946.1</td>
      <td>689;6692</td>
      <td>thrA;Not_found</td>
      <td>UTI89_RS00010;UTI89_RS00040</td>
      <td>+;-</td>
      <td>-;-</td>
      <td>0.14372716199756394;0.8057302585604472</td>
      <td>True;True</td>
      <td>Unambiguous;Ambiguous</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>A_1_PC01_PR03</th>
      <td>A_1_PC01_PR03</td>
      <td>A</td>
      <td>1</td>
      <td>PC01</td>
      <td>PR03</td>
      <td>NC_007946.1;NC_007946.1;NC_007946.1</td>
      <td>4056;5156;5581</td>
      <td>thrC;nan;yaaA</td>
      <td>UTI89_RS00020;nan;UTI89_RS00035</td>
      <td>+;nan;-</td>
      <td>+;-;+</td>
      <td>0.2517482517482518;nan;0.9832689832689833</td>
      <td>True;False;True</td>
      <td>Ambiguous;Ambiguous;Ambiguous</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>A_1_PC01_PR04</th>
      <td>A_1_PC01_PR04</td>
      <td>A</td>
      <td>1</td>
      <td>PC01</td>
      <td>PR04</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>A_1_PC01_PR05</th>
      <td>A_1_PC01_PR05</td>
      <td>A</td>
      <td>1</td>
      <td>PC01</td>
      <td>PR05</td>
      <td>NC_007946.1</td>
      <td>5715</td>
      <td>yaaA</td>
      <td>UTI89_RS00035</td>
      <td>-</td>
      <td>-</td>
      <td>0.810811</td>
      <td>True</td>
      <td>Ambiguous</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>H_12_PC08_PR04</th>
      <td>H_12_PC08_PR04</td>
      <td>H</td>
      <td>12</td>
      <td>PC08</td>
      <td>PR04</td>
      <td>NC_007946.1;NC_007946.1</td>
      <td>3370;3884</td>
      <td>thrB;thrC</td>
      <td>UTI89_RS00015;UTI89_RS00020</td>
      <td>+;+</td>
      <td>+;-</td>
      <td>0.6120042872454448;0.11810411810411811</td>
      <td>True;True</td>
      <td>Unambiguous;Ambiguous</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>H_12_PC08_PR05</th>
      <td>H_12_PC08_PR05</td>
      <td>H</td>
      <td>12</td>
      <td>PC08</td>
      <td>PR05</td>
      <td>NC_007946.1</td>
      <td>3566</td>
      <td>thrB</td>
      <td>UTI89_RS00015</td>
      <td>+</td>
      <td>+</td>
      <td>0.822079</td>
      <td>True</td>
      <td>Ambiguous</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>H_12_PC08_PR06</th>
      <td>H_12_PC08_PR06</td>
      <td>H</td>
      <td>12</td>
      <td>PC08</td>
      <td>PR06</td>
      <td>NC_007946.1</td>
      <td>5399</td>
      <td>Not_found</td>
      <td>UTI89_RS00030</td>
      <td>+</td>
      <td>+</td>
      <td>0.56229</td>
      <td>True</td>
      <td>Ambiguous</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>H_12_PC08_PR07</th>
      <td>H_12_PC08_PR07</td>
      <td>H</td>
      <td>12</td>
      <td>PC08</td>
      <td>PR07</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
    </tr>
    <tr>
      <th>H_12_PC08_PR08</th>
      <td>H_12_PC08_PR08</td>
      <td>H</td>
      <td>12</td>
      <td>PC08</td>
      <td>PR08</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
      <td>nan</td>
    </tr>
  </tbody>
</table>
<p>6144 rows × 15 columns</p>
</div></div>
</div>
</section>
<section id="References">
<h2>References<a class="headerlink" href="#References" title="Permalink to this headline"></a></h2>
<p>[1] Langmead, B. and Salzberg, S.L., 2012. Fast gapped-read alignment with Bowtie 2. Nature methods, 9(4), pp.357-359.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="run_test_data_cmdline.html" class="btn btn-neutral float-right" title="How to run arraylib-solve on the command line" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Tuemay Capraz.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>